[54461]: Query block toggle - get all descendants of current page

PaulREnglish opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- Link this PR to its associated issue with an appropriate keyword: Closes, See, Follow up to, etc. -->
Resolves: https://github.com/WordPress/gutenberg/issues/54461

<!-- In a few words, what is the PR actually doing? -->
Adds a toggle to set the current page as the parent page. 

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
See https://github.com/WordPress/gutenberg/issues/54461#issuecomment-2608989671.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
Switching the 'Set parent as current page' toggle to the 'on' position will set the `Parents` query to an array containing only the current page's id. Switching this toggle off will empty this array. 

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
1. Open a page (or any other hierarchal post type) in the editor
2. Insert a query block
3. Open the filters dropdown (click the 3 dots next to 'Filters' in the sidebar) and insert a 'Parents' filter
4. Click 'Set parent as current page' to set the current page as the only parent.
5. Child pages of the current page should be displayed in the query block
6. Click save and go to front end
7. Child pages of the current page should be displayed in the query block on the frontend

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
1. In the page editor, insert a query block
2. Using the tab keys navigate to the 'Set parent as current page' toggle
3. Use the space bar to toggle

## Screenshots or screencast <!-- if applicable -->

<!-- If you would like to upload screenshots, feel free to use the table below when it is useful to show the difference between before and after the change. -->

|Before|After|
|-|-|
|![Screenshot on 2025-02-20 at 14-52-11](https://github.com/user-attachments/assets/7b798a68-0318-44bb-8055-1c6ec0dc95e8)|![Screenshot on 2025-02-20 at 14-51-23](https://github.com/user-attachments/assets/04817fc1-3d4c-478b-ae66-cfa88ee6e549)|

PaulREnglish committed:
feat(query-block): add 'set parent as current page' toggle

PaulREnglish committed:
Merge branch 'WordPress:trunk' into feature/54461-query-block-parent-child-toggles

PaulREnglish committed:
feat(query-block): add show all descendants toggle

PaulREnglish committed:
refactor(query-block): hide descendents toggle when not needed

PaulREnglish committed:
refactor(query-block): remove limit in sql query when ancestor is set

To get descendants of a particular page, we first need to fetch all pages so we can pass them through get_page_children()

PaulREnglish commented:
Here's a demo of how it currently works.  Still a WIP but any feedback would be appreciated:

https://github.com/user-attachments/assets/4083be77-e0d5-4412-9a0f-e878f00483d3

Drivingralle commented:
What do you think about add the setting to the existing input field like this?

![408529130-6f5f9552-3c72-42eb-b919-b2677ef25cc6](https://github.com/user-attachments/assets/28404499-e7e8-4d3b-a3f2-00ff6f10d14c)

PaulREnglish commented:
> What do you think about add the setting to the existing input field like this?
> 
> ![408529130-6f5f9552-3c72-42eb-b919-b2677ef25cc6](https://private-user-images.githubusercontent.com/1727269/410385818-28404499-e7e8-4d3b-a3f2-00ff6f10d14c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg4MzQ0MzEsIm5iZiI6MTczODgzNDEzMSwicGF0aCI6Ii8xNzI3MjY5LzQxMDM4NTgxOC0yODQwNDQ5OS1lN2U4LTRkM2ItYTNmMi0wMGZmNmYxMGQxNGMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDIwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAyMDZUMDkyODUxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmUzNjI1ODdhMDBkZDk0ZWM5MjBmZWIxMTE1ODJkNTYwMjYxZTE4NWIzMWQyNTZmNWJjM2U5OTBjYWU2ZjE0ZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.qe3Y29NfaksOl9UlnTx_EIZ9nPFCAcIMh3retqUQ7x4)

Hi :wave:!
During initial implementation, I had made the toggle set the existing text/input field to contain just the current page without hiding it like you suggested.  However, the second toggle (the "Show all descendants one") will only work if we have exactly one parent (as [get_page_children](https://developer.wordpress.org/reference/functions/get_page_children/) accepts one page id).  So to avoid editors using the show all descendants toggle on multiple parents, I hid the text/input field when that toggle is visible.  Hence, the current behaviour shown in my demo.

PaulREnglish commented:
Does anyone have any thoughts regarding the approach I've taken codewise with regards to getting all descendants?  
Currently the query fetches _all_ pages from the database before filtering them with the use of [get_page_children](https://developer.wordpress.org/reference/functions/get_page_children/).  This has required the use of 3 filters (`query_loop_block_query_vars`, `post_limits_request` and `posts_results`) and the use of a new `post_ancestor` attribute.

An alternative approach I did consider but haven't used would be to loop over the page hierarchy and execute a different sql query for every depth we need to fetch children from. (So one sql query would fetch the direct children of the parent, a second sql query would fetch the children of those children, etc, etc).

PaulREnglish committed:
fix(query-block): make ancestor attribute work correctly in editor

PaulREnglish commented:
After some discussion in Slack, I'm going to try to add the `post_ancestor` query property to WordPress core so I don't have to use the 'hacky' approach of using filters.

See ticket here: https://core.trac.wordpress.org/ticket/62953#ticket

Mamaduka commented:
Thanks again for working on this, @PaulREnglish!

Let's continue the technical discussion around `post_ancestor` in the core ticket. Once that's handled, the UI can be updated for the Query block.

I'm going to remove this enhancement request from the WP 6.8 project board.

PaulREnglish commented:
> Thanks again for working on this, @PaulREnglish!

My pleasure!

> Let's continue the technical discussion around `post_ancestor` in the core ticket. Once that's handled, the UI can be updated for the Query block.
> 
> I'm going to remove this enhancement request from the WP 6.8 project board.

Sounds good.

joemcgill commented:
@PaulREnglish raised this PR and [the related Trac ticket](https://core.trac.wordpress.org/ticket/62953) in today's Dev Chat. While I agree that adding core support for querying all ancestors will require some discussion that will cause this to miss the 6.8 milestone, I think it might be worth reducing the scope of this PR to only address the original issue, which is to set the query block to show only direct children of the current page.

I think that could be done with existing functionality and then iterated on in future versions of GB that would be included in 6.9.

PaulREnglish committed:
revert(query): revert ancestor prop implementation using filters

The usage of filters was a rather hacky implementation as the ancestor query prop ought to be implemented in WordPress Core rather than Gutenberg.

PaulREnglish commented:
> @PaulREnglish raised this PR and [the related Trac ticket](https://core.trac.wordpress.org/ticket/62953) in today's Dev Chat. While I agree that adding core support for querying all ancestors will require some discussion that will cause this to miss the 6.8 milestone, I think it might be worth reducing the scope of this PR to only address the original issue, which is to set the query block to show only direct children of the current page.
> 
> I think that could be done with existing functionality and then iterated on in future versions of GB that would be included in 6.9.

Thanks very much for your input!  I've reverted the ancestor/show all descendants toggle leaving behind only the "
Set parent as current page" toggle which should hopefully address the original issue.

I'll record a video of how it works then I can mark this PR as ready to review.
